### PR TITLE
Fix upwind/downwind finite volume operators

### LINKED ISSUE #3526: [Bug]: Upwind/downwind schemes are not working properly
### PyBaMM Version

develop

### Python Version

all

### Describe the bug

Checking the problem solved in `finite-volumes.ipynb` it does not correspond with what one would expect from an upwind scheme.

See example below.

### Steps to Reproduce
The example below compares a toy problem with its analytical solution. Note that for coarse meshes there is some disagreement but it converges if you refine the mesh. Commenting the lines that fix the system retrieves the previous behaviour which disagrees with the analytical solution.

```python3
import pybamm


model = pybamm.BaseModel()
x = pybamm.SpatialVariable("x", domain="domain", coord_sys="cartesian")
u = pybamm.Variable("u", domain="domain")
u_an = x + (pybamm.t - x) * ((x - pybamm.t) > 0)
v = pybamm.PrimaryBroadcastToEdges(1, ["domain"])
rhs = - pybamm.div(pybamm.upwind(u) * v) + 1
model.boundary_conditions = {
    u: {
        "left": (pybamm.Scalar(0), "Dirichlet"),
    }
}
model.rhs = {u: rhs}
model.initial_conditions = {u: pybamm.Scalar(0)}
model.variables = {"u": u, "x": x, "analytical": u_an}
geometry = {"domain": {x: {"min": pybamm.Scalar(0), "max": pybamm.Scalar(1)}}}
submesh_types = {"domain": pybamm.Uniform1DSubMesh}
N = 5
var_pts = {x: N}
mesh = pybamm.Mesh(geometry, submesh_types, var_pts)
spatial_methods = {"domain": pybamm.FiniteVolume()}
disc = pybamm.Discretisation(mesh, spatial_methods)
model_disc = disc.process_model(model, inplace=False)

# Fix the entries of the discretised system (uncomment to get original behaviour)
for i in range(N - 1):
    model_disc.rhs[u].children[1].children[0].entries[i + 1, i] = 0
    model_disc.rhs[u].children[1].children[0].entries[i + 1, i + 1] = -N
    model_disc.rhs[u].children[1].children[0].entries[i + 1, i + 2] = N
model_disc.rhs[u].children[1].children[0].entries[0, 0] = - 2 * N
model_disc.rhs[u].children[1].children[0].entries[0, 1] = 2 * N


print(model_disc.rhs[u].children[1].children[0].entries.toarray())
solver = pybamm.CasadiSolver()
solution =solver.solve(model_disc, [0, 10])

plot = pybamm.QuickPlot(solution,[["u", "analytical"]])
plot.dynamic_plot()
```

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by brosaplanella**
# Description

Fixed the issue with the upwind and downwind operators. I am now leveraging the `add_ghost_nodes` method, to which I had to do some minor changes.

Fixes #3526 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: tests/unit/test_spatial_methods/test_finite_volume/test_finite_volume.py [hunk 1]
```diff
@@ -408,13 +408,14 @@ def test_upwind_downwind(self):
         self.assertEqual(disc_downwind.size, nodes.size + 1)
 
         y_test = 2 * np.ones_like(nodes)
+        print(disc_upwind.evaluate(y=y_test))
```
Comment (tinosulzer) [inline tests/unit/test_spatial_methods/test_finite_volume/test_finite_volume.py:411]: can this be removed?

**Review by tinosulzer — APPROVED** [review]
Thanks, looks good. Can you show plots of the upwinding example vs analytical solution before and after the fix?

**Comment by brosaplanella** [comment]
# Before & after plots
$$ u_t + u_x = s, \quad 0 < x  < 1$$

$$u(t,0) = b$$

$$u(0,x) = 0$$

## Example 1
s = 1, b = 0

### Before



### After



## Example 2
s = 0, b = 1

### Before


### After

**Comment by valentinsulzer** [comment]
Thanks, that does look much better! Is the mismatch with the analytical solution because it's not a high-enough order scheme?

**Comment by brosaplanella** [comment]
I think so, as refining the mesh gives a better result. I need to do some more reading on this, and potentially implement higher order methods, but that's a separate issue.

**Pull request merged**